### PR TITLE
[SPARK-19338][SQL] Add UDF names in explain

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added a variable for a UDF name in `ScalaUDF`.
Then, if the variable filled, `DataFrame#explain` prints the name.

## How was this patch tested?
Added a test in `UDFSuite`.

### CODE REVIEW

**Comment by rxin** [comment]
Maybe add a prefix so it is clear a UDF? e.g. `UDF:func_name(...)`

**Comment by maropu** [comment]
Aha, SGTM. I'll fix

**Comment by rxin** [comment]
LGTM pending jenkins.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 1]
```diff
@@ -125,7 +125,7 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
   def register[RT: TypeTag](name: String, func: Function0[RT]): UserDefinedFunction = {
     val dataType = ScalaReflection.schemaFor[RT].dataType
     val inputTypes = Try(Nil).toOption
-    def builder(e: Seq[Expression]) = ScalaUDF(func, dataType, e, inputTypes.getOrElse(Nil))
+    def builder(e: Seq[Expression]) = ScalaUDF(func, dataType, e, inputTypes.getOrElse(Nil), Some(name))
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: Oh, I remember I was told that we should remove or fix the comments above to generate this codes too.

Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: oh, yea, you're right. Thanks! I'll fix

**Comment by maropu** [comment]
It seems the latest test failure does not relate to this pr..., I'll test again.

**Review by gatorsmile — APPROVED** [review]
LGTM pending jenkins.

**Comment by maropu** [comment]
Jenkins, retest this please.

**Comment by maropu** [comment]
@gatorsmile @rxin okay, it seems ready to merge.

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Pull request closed without merging**